operator/internal/controller/pulumi/program_controller_test.go

```diff
@@ -0,0 +1,159 @@
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	v1 "github.com/pulumi/pulumi-kubernetes-operator/operator/api/pulumi/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ = Describe("Program Controller", func() {
+	var (
+		program               v1.Program
+		r                     *ProgramReconciler
+		programNamespacedName types.NamespacedName
+		ctx                   context.Context
+		advertisedAddress     string
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+
+		advertisedAddress = "http://fake-svc.fake-namespace.svc.cluster.local"
+
+		r = &ProgramReconciler{
+			Client:   k8sClient,
+			Scheme:   k8sClient.Scheme(),
+			Recorder: record.NewFakeRecorder(10),
+			ProgramHandler: &ProgramHandler{
+				k8sClient: k8sClient,
+				address:   advertisedAddress,
+			},
+		}
+
+		programNamespacedName = types.NamespacedName{
+			Name:      fmt.Sprintf("test-program-%s", utilrand.String(8)),
+			Namespace: "default",
+		}
+
+		program = v1.Program{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       programNamespacedName.Name,
+				Namespace:  programNamespacedName.Namespace,
+				Generation: 1,
+			},
+			Program: v1.ProgramSpec{
+				Resources: map[string]v1.Resource{
+					"test-resource": {
+						Type: "kubernetes:core/v1:Pod",
+					},
+				},
+			},
+		}
+		Expect(k8sClient.Create(ctx, &program)).Should(Succeed())
+	})
+
+	AfterEach(func() {
+		Expect(k8sClient.Delete(ctx, &program)).Should(Succeed())
+	})
+
+	reconcileFn := func(ctx context.Context) (reconcile.Result, error) {
+		return r.Reconcile(ctx, reconcile.Request{NamespacedName: programNamespacedName})
+	}
+
+	When("reconciling a resource", func() {
+		It("should generate an artifact URL in the status", func() {
+			By("not expecting a status to be present before first reconciliation")
+			program := v1.Program{}
+			Expect(k8sClient.Get(ctx, programNamespacedName, &program)).Should(Succeed())
+			Expect(program.Status.Artifact).To(BeNil())
+			Expect(program.GetGeneration()).To(Equal(int64(1)))
+
+			By("reconciling a new Program object")
+			_, err := reconcileFn(ctx)
+			Expect(err).NotTo(HaveOccurred())
+
+			program = v1.Program{}
+			Expect(k8sClient.Get(ctx, programNamespacedName, &program)).Should(Succeed())
+			Expect(program.Status.Artifact).NotTo(BeNil())
+			Expect(program.Status.ObservedGeneration).To(Equal(int64(1)))
+			Expect(program.Status.Artifact.URL).
+				To(Equal(fmt.Sprintf("%s/programs/%s/%s", advertisedAddress, programNamespacedName.Namespace, programNamespacedName.Name)))
+
+			By("reconciling the same Program object but with a new spec change")
+			program.Program.Resources = map[string]v1.Resource{
+				"test-resource": {
+					Type: "kubernetes:core/v1:Service",
+				},
+			}
+			program.ObjectMeta.Generation++
```

Comment:
Why are you incrementing the generation manually? The Update call itself will do that.